internal/controllers/utils/utils.go

```diff
@@ -771,6 +772,68 @@ func CheckClusterLabelsForPolicies(
 	return nil
 }
 
+// FindClusterInstanceImmutableFieldUpdates identifies updates made to immutable fields
+// in the ClusterInstance spec. It returns a list of paths for the updated fields that
+// are considered immutable and should not be modified.
+func FindClusterInstanceImmutableFieldUpdates(
+	oldClusterInstance, newClusterInstance *unstructured.Unstructured) ([]string, error) {
+
+	// Non-immutable ClusterInstance fields at the cluster-level
+	allowedClusterFields := []string{
+		"extraAnnotations",
+		"extraLabels",
+	}
+	// Non-immutable ClusterInstance fields at the node-level
+	allowedNodeFields := []string{
+		"extraAnnotations",
+		"extraLabels",
+	}
+
+	oldClusterInstanceSpec := oldClusterInstance.Object["spec"].(map[string]any)
+	newClusterInstanceSpec := newClusterInstance.Object["spec"].(map[string]any)
+
+	diffs, err := diff.Diff(oldClusterInstanceSpec, newClusterInstanceSpec)
+	if err != nil {
+		return nil, fmt.Errorf("error comparing differences between existing "+
+			"and newly rendered ClusterInstance: %w", err)
+	}
+
+	var updatedFields []string
+	for _, diff := range diffs {
+		/* Examples of diff result in json format
+
+		Label added at the cluster-level
+		  {"type": "create", "path": ["extraLabels", "ManagedCluster", "newLabelKey"], "from": null, "to": "newLabelValue"}
+
+		Field updated at the cluster-level
+		  {"type": "update", "path": ["baseDomain"], "from": "domain.example.com", "to": "newdomain.example.com"}
+
+		New node added
+		  {"type": "create", "path": ["nodes", "1"], "from": null, "to": {"hostName": "worker2"}}
+
+		Field updated at the node-level
+		  {"type": "update", "path": ["nodes", "0", "nodeNetwork", "config", "dns-resolver", "config", "server", "0"], "from": "192.10.1.2", "to": "192.10.1.3"}
+		*/
+
+		oranUtilsLog.Info(
+			fmt.Sprintf(
+				"Detected field change in the newly rendered ClusterInstance(%s) type: %s, path: %s, from: %+v, to: %+v",
+				oldClusterInstance.GetName(), diff.Type, strings.Join(diff.Path, "."), diff.From, diff.To,
+			),
+		)
+
+		if diff.Path[0] == "nodes" {
+			if len(diff.Path) == 2 || (len(diff.Path) > 2 && !slices.Contains(allowedNodeFields, diff.Path[2])) {
```

Comment:
Can you double-check the logic here?  It seems odd that if the `len(diff.Path) == 2` that we are assuming that the field should be added to `updatedFields` regardless of its name.

Reply:
Yes, in the case of len(diff.Path) == 2, a new node is being added to the ClusterInstance, and it must contain immutable fields as they are mandatory in the template.

However, I do want to update something else—we should allow new nodes to be added after successful cluster provisioning (even though we’re not officially considering scaling in/out at the moment).

Reply:
ok